### PR TITLE
JSON options and validation

### PR DESCRIPTION
**Pull request by jeffeaton**
* Edits to JSON model options
* Add `validate_model_options(...)` call in `hintr_run_model(...)` so that invalid inputs will be flagged with useful error message upon Run Model click.

I added an additional step to the progress printing. Confirm that doesn't have broader consequences. If so, then remove that.

Thanks,
Jeff

### CODE REVIEW

**Comment by richfitz** [comment]
See https://github.com/mrc-ide/hintr/tree/mrc-1166 and associated builds

**Comment by r-ash** [comment]
You can use an online json validator to check your json is valid.

Something like https://jsonformatter.curiousconcept.com/

**Comment by richfitz** [comment]
Is the error coming from one of the ones that gets interpolated though?

**Comment by jeffeaton** [comment]
Thank you for pointing me towards the hintr tests. I had misunderstood that the naomi tests were checking the JSON (and indeed a bit surprised that I wasn't getting more errors in my JSON updates...).

I've added three more commits to this PR to fix the JSON errors and have made a PR on hintr to update the tests required by the model options changes: https://github.com/mrc-ide/hintr/pull/93

Thanks,
Jeff

File: R/run-model.R [hunk 1]
```diff
@@ -47,13 +45,29 @@ hintr_run_model <- function(data, options, output_path = tempfile(),
 
   INLA:::inla.dynload.workaround()
   progress <- new_progress()
+
+  progress$start("Validating inputs and options")
```
Comment (r-ash) [inline R/run-model.R]: Nice!

File: inst/metadata/art_run_options.json [hunk 2]
```diff
@@ -3,25 +3,76 @@
...
+		  "required": true,
+		  "options":
+		  [
+		      {
+			  "id": "false",
+			  "label": "No"
+		      },
+		      {
+			  "id": "true",
+			  "label": "yes"
```
Comment (r-ash) [inline inst/metadata/art_run_options.json]: Let's use a capital letter to be consistent.

File: inst/metadata/art_run_options.json [hunk 3]
```diff
@@ -3,25 +3,76 @@
...
+		  "value": "false",
+		  "helpText": "Model proportion of ART residents who seek treatment in neighboring districts."
+	      }
+	  ]
+      },
+      {
+	  "label": "Logit prior mean for neighboring district attraction.",
+	  "controls": [
+	      {
+		  "name": "log_or_gamma_offset",
```
Comment (r-ash) [inline inst/metadata/art_run_options.json]: Is this used anywhere? Planned to be added?

File: tests/testthat/test-model-options.R [hunk 4]
```diff
@@ -54,14 +50,51 @@ test_that("validate model options returns true", {
...
+    anc_prevalence_year1 = 2016,
+    anc_prevalence_year2 = 2018,
+    anc_art_coverage_year1 = 2016,
+    anc_art_coverage_year2 = 2018,
+    no_of_samples = 20
   )
   expect_true(validate_model_options(data, options))
 })
 
+test_that("validate model options returns error for invalid", {
```
Comment (r-ash) [inline tests/testthat/test-model-options.R]: Nice test

**Comment by r-ash** [comment]
I'm also seeing some warnings when running the model on this branch. Are they important?
```
Warning messages:
1: Column `artdat_sex` joining factor and character vector, coercing into character vector 
2: Column `area_id` joining factor and character vector, coercing into character vector 
3: Column `artdat_sex` joining factor and character vector, coercing into character vector 
4: Column `area_id` joining factor and character vector, coercing into character vector 
```

File: inst/metadata/art_run_options.json (hunk 2 above)
Comment (r-ash) [reply inst/metadata/art_run_options.json]: I've fixed this

**Comment by r-ash** [comment]
Is this what you expect? Empty output options section?
**Survey only**
Selection_001
Selection_002

**Survey, ANC and ART**
Selection_003
Selection_004
Selection_005

**Comment by jeffeaton** [comment]
I think that these are addressed in #40. Should raise issue if they still appear after merging that one.

Thanks,
Jeff

On Dec 2, 2019, at 22:21, Rob <notifications@github.com> wrote:

﻿

I'm also seeing some warnings when running the model on this branch. Are they important?

Warning messages:
1: Column `artdat_sex` joining factor and character vector, coercing into character vector
2: Column `area_id` joining factor and character vector, coercing into character vector
3: Column `artdat_sex` joining factor and character vector, coercing into character vector
4: Column `area_id` joining factor and character vector, coercing into character vector

**Pull request merged**